woocommerce/woocommerce-gutenberg-products-block
# Update WordPress Monorepo packages with React 17x support

This PR exists to update all WordPress dependencies to the latest and greatest versions, and since they require React 17x, fixes conflicts with other dependencies we do not directly manage.

Fixes #4533
Closes #4904

Work involved:

- Updated all `@wordpress/` dependencies, including those raised in #4904 which have been Blocked by the React 17x requirement
- Audited all dependencies for React support using ` npm ls react`
- Moved `mapped-types` back from `@automattic/data-stores` since having that as a dependency for 3 functions seems too much. `@automattic/data-stores` has not been updated for 8 months and does not support React 17x. There are some other versions of `@automattic/data-stores` within the org with React 17x support but these are not deployed to NPM. It's unclear where the single source of truth is, so to unblock this work I've just moved the types across.
- Moved `SearchListControl` back into our `editor-components`, from `@woocommerce/components` to remove that dependency. `@woocommerce/components` is another package that does not support React 17x (I will log this), but again, to unblock us, and prevent future conflicts, I spoke to @senadir and moved this back into our repo. I don't think we can just widen the React version requirements in `@woocommerce/components` to fix this because there are other dependencies (from the WordPress monorepo) that also need bumping, similar to the work in this PR.
- `SearchListControl` uses now-deprecated HOCs such as `withState`, so I converted this to Typescript and functional components with hooks to resolve. This is again something I could contribute back to `@woocommerce/components`, but until the dependency issues are resolved and monorepos updated, I think this will be problematic. It's not clear to me who is managing this type of work (for us it's porter and we seem more active with dependency updates).

### Testing

Developer Smoke Testing:

1. Npm install successfully
2. `npm ls react` should only list versions of React 17x, not 16x
3. Successful Build
4. Successful Tests
5. Go to the editor and smoke test block insertion. Check for console errors.
6. Go to frontend and smoke test block rendering. Check for console errors.

`SearchListControl`:

1. Insert a product block - test Products by Attribute and Handpicked Products
2. Confirm you can toggle and check items in the search list
3. Confirm you can remove selected items using the tag inputs above the list
4. Confirm you can search for an item and the list is filtered accordingly
5. Confirm you can save and results persist

Aside from smoke testing, there are no user facing changes to test here.

### Performance Impact

Should not be impacted. With fewer dependencies, if anything,  builds and install should be slightly faster. App performance should not be impacted.
